README.md
# bedrock-permission-http
 Bedrock permission management via http endpoints


CHANGELOG.md
# bedrock-permission-http ChangeLog

## 1.1.0 - 2020-07-10

## Changed
- Update peer dependencies.

## 1.0.3 - 2016-11-28

### Fixed
- Fix the DELETE API.

## 1.0.2 - 2016-06-07

### Changed
- Update dependencies.

## 1.0.1 - 2016-05-03

### Fixed
- Fixed input validation in the session.get event handler.

## 1.0.0 - 2016-04-28

- See git history for changes.


lib/main.js
/*
 * Copyright (c) 2012-2016 Digital Bazaar, Inc. All rights reserved.
 */
/* jshint node: true */
'use strict';

var _ = require('lodash');
var bedrock = require('bedrock');
var BedrockError = bedrock.util.BedrockError;
var config = bedrock.config;
var brPermission = require('bedrock-permission');
var brPassport = require('bedrock-passport');
var database = require('bedrock-mongodb');
var ensureAuthenticated = brPassport.ensureAuthenticated;
var validate = require('bedrock-validation').validate;

require('bedrock-express');

// load config
require('./config');

bedrock.events.on('bedrock.configure', function() {
  config.permission.roleBaseUrl =
    config.server.baseUri + config['permission-http'].routes.roles;
});

bedrock.events.on('bedrock-express.configure.routes', function(app) {
  var permissionsBasePath = config['permission-http'].routes.permissions;
  var permissionsBaseUrl = config.server.baseUri + permissionsBasePath;
  var rolesBasePath = config['permission-http'].routes.roles;

  /**
   * Retrieves all permissions from the config held permissions.
   */
  app.get(permissionsBasePath, ensureAuthenticated, function(req, res, next) {
    brPermission.getPermissions(req.user.identity, function(err, permissions) {
      if(err) {
        return next(err);
      }
      res.json(permissions);
    });
  });

  /**
   * Gets single role
   */
  app.get(
    rolesBasePath + '/:id', ensureAuthenticated, function(req, res, next) {
    var roleId = req.params.id;
    if(config.permission.roleBaseUrl.length !== 0) {
      roleId = config.permission.roleBaseUrl + '/' + roleId;
    }
    brPermission.getRole(req.user.identity, roleId, function(err, role) {
      if(err) {
        return next(err);
      }
      res.json(role);
    });
  });

  /**
   * Gets all existing roles
   */
  app.get(rolesBasePath, ensureAuthenticated, function(req, res, next) {
    brPermission.getRoles(req.user.identity, function(err, roles) {
      if(err) {
        return next(err);
      }
      res.json(roles);
    });
  });

  /**
   * Inserts a newly created role to the database
   */
  app.post(
    rolesBasePath, ensureAuthenticated, validate('services.role.postRole'),
    function(req, res, next) {
    brPermission.addRole(req.user.identity, req.body, function(err, result) {
      if(err) {
        if(database.isDuplicateError(err)) {
          return next(new BedrockError(
            'Duplicate role.', 'DuplicateRole', {
            httpStatusCode: 409,
            public: true
          }));
        }
        return next(err);
      }
      res.status(201).json(result.role);
    });
  });

  /**
   * Edits a role that is already in the database
   */
  app.patch(
    rolesBasePath + '/:id', ensureAuthenticated,
    validate('services.role.patchRole'), function(req, res, next) {
    var updateOperations = req.body.filter(function(o) {
      return o.op == 'updateRole';
    }).map(function(o) {
      return o.changes;
    });
    if(updateOperations.length !== req.body.length) {
      return next(new BedrockError(
        'Operations not implemented.', 'NotImplemented',
        {httpStatusCode: 400, 'public': true}
      ));
    }
    // the validation schema specifies that the must be only one operation
    brPermission.updateRole(
      req.user.identity, updateOperations[0], function(err, result) {
      if(err) {
        return next(err);
      }
      res.json(result);
    });
  });

  /**
   * Deletes a role with the specified ID
   */
  app.delete(
    rolesBasePath + '/:id', ensureAuthenticated, function(req, res, next) {
    var roleId = req.params.id;
    brPermission.removeRole(req.user.identity, roleId, function(err, result) {
      if(err) {
        return next(err);
      }
      res.sendStatus(204);
    });
  });

}); // end configure routes

// add a permission table to the session data
bedrock.events.on(
  'bedrock-session-http.session.get', function(req, session, callback) {
  if(req.isAuthenticated() && req.user.identity && req.user.identity.id) {
    session.identity = session.identity || {};
    session.identity.sysResourceRole = req.user.identity.sysResourceRole;
    return brPermission.createPermissionTable(
      session.identity.sysResourceRole, function(err, table) {
      if(err) {
        return callback(err);
      }
      session.identity.sysPermissionTable = table;
      callback();
    });
  }
  callback();
});


schemas/services.role.js
/*
 * Copyright (c) 2016-2020 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const schemas = require('bedrock-validation').schemas;

const postRole = {
  type: 'object',
  properties: {
    comment: schemas.description({required: false}),
    label: schemas.label({required: false}),
    sysPermission: {
      type: 'array',
      required: true,
      minItems: 1,
      uniqueItems: true,
      items: {
        type: 'string'
      }
    }
  },
  additionalProperties: false
};

const patchRole = {
  type: 'array',
  required: true,
  minItems: 1,
  maxItems: 1,
  items: {
    type: 'object',
    properties: {
      op: {type: 'string', required: true},
      changes: {
        id: schemas.identifier(),
        comment: schemas.description({required: false}),
        label: schemas.label({required: false}),
        sysPermission: {
          type: 'array',
          required: true,
          minItems: 1,
          uniqueItems: true,
          items: {
            type: 'string'
          }
        }
      }
    },
    additionalProperties: false
  }
};

module.exports.postRole = function() {
  return postRole;
};

module.exports.patchRole = function() {
  return patchRole;
};


lib/index.js
/*
 * Copyright (c) 2012-2020 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const bedrock = require('bedrock');
const BedrockError = bedrock.util.BedrockError;
const config = bedrock.config;
const brPermission = require('bedrock-permission');
const brPassport = require('bedrock-passport');
const database = require('bedrock-mongodb');
const ensureAuthenticated = brPassport.ensureAuthenticated;
const util = require('util');
const validate = require('bedrock-validation').validate;

require('bedrock-express');

// load config
require('./config');

bedrock.events.on('bedrock.configure', function() {
  config.permission.roleBaseUrl =
    config.server.baseUri + config['permission-http'].routes.roles;
});

bedrock.events.on('bedrock-express.configure.routes', function(app) {
  const permissionsBasePath = config['permission-http'].routes.permissions;
  const rolesBasePath = config['permission-http'].routes.roles;

  /**
   * Retrieves all permissions from the config held permissions.
   */
  app.get(permissionsBasePath, ensureAuthenticated, function(req, res, next) {
    brPermission.getPermissions(req.user.identity, function(err, permissions) {
      if(err) {
        return next(err);
      }
      res.json(permissions);
    });
  });

  /**
   * Gets single role
   */
  app.get(
    rolesBasePath + '/:id', ensureAuthenticated, function(req, res, next) {
      let roleId = req.params.id;
      if(config.permission.roleBaseUrl.length !== 0) {
        roleId = config.permission.roleBaseUrl + '/' + roleId;
      }
      brPermission.getRole(req.user.identity, roleId, function(err, role) {
        if(err) {
          return next(err);
        }
        res.json(role);
      });
    });

  /**
   * Gets all existing roles
   */
  app.get(rolesBasePath, ensureAuthenticated, function(req, res, next) {
    brPermission.getRoles(req.user.identity, function(err, roles) {
      if(err) {
        return next(err);
      }
      res.json(roles);
    });
  });

  /**
   * Inserts a newly created role to the database
   */
  app.post(
    rolesBasePath, ensureAuthenticated, validate('services.role.postRole'),
    function(req, res, next) {
      brPermission.addRole(req.user.identity, req.body, function(err, result) {
        if(err) {
          if(database.isDuplicateError(err)) {
            return next(new BedrockError(
              'Duplicate role.', 'DuplicateRole', {
                httpStatusCode: 409,
                public: true
              }));
          }
          return next(err);
        }
        res.status(201).json(result.role);
      });
    });

  /**
   * Edits a role that is already in the database
   */
  app.patch(
    rolesBasePath + '/:id', ensureAuthenticated,
    validate('services.role.patchRole'), function(req, res, next) {
      const updateOperations = req.body.filter(function(o) {
        return o.op == 'updateRole';
      }).map(function(o) {
        return o.changes;
      });
      if(updateOperations.length !== req.body.length) {
        return next(new BedrockError(
          'Operations not implemented.', 'NotImplemented',
          {httpStatusCode: 400, public: true}
        ));
      }
      // the validation schema specifies that the must be only one operation
      brPermission.updateRole(
        req.user.identity, updateOperations[0], function(err, result) {
          if(err) {
            return next(err);
          }
          res.json(result);
        });
    });

  /**
   * Deletes a role with the specified ID
   */
  app.delete(
    rolesBasePath + '/:id', ensureAuthenticated, function(req, res, next) {
      const roleId = createRoleId(req.params.id);
      brPermission.removeRole(req.user.identity, roleId, function(err) {
        if(err) {
          return next(err);
        }
        res.sendStatus(204);
      });
    });

  /**
   * Creates a Role ID URL from the given ID.
   *
   * @param id the ID to use.
   *
   * @return the role ID URL for the role.
   */
  function createRoleId(id) {
    return util.format(
      '%s%s/%s', bedrock.config.server.baseUri, rolesBasePath,
      encodeURIComponent(id));
  }
}); // end configure routes

// add a permission table to the session data
bedrock.events.on(
  'bedrock-session-http.session.get', function(req, session, callback) {
    if(req.isAuthenticated() && req.user.identity &&
      req.user.identity.sysResourceRole) {
      session.identity = session.identity || {};
      session.identity.sysResourceRole = req.user.identity.sysResourceRole;
      return brPermission.createPermissionTable(
        session.identity.sysResourceRole, function(err, table) {
          if(err) {
            return callback(err);
          }
          session.identity.sysPermissionTable = table;
          callback();
        });
    }
    callback();
  });
